### PR TITLE
Enable batching rule for `debug_print`

### PR DESCRIPTION
**Pull request by sharadmv**
The batching rule for `debug_print` unrolls the callback across the mapped axis to avoid enforcing rank-polymorphism for the callback and print.

### CODE REVIEW

File: jax/_src/debugging.py [hunk 1]
```diff
@@ -48,11 +53,21 @@ def debug_callback_abstract_eval(*flat_avals, callback: Callable[..., Any],
...
+  """Unrolls the debug callback across the mapped axis."""
+  axis_size = next(x.shape[i] for x, i in zip(args, dims)
+                   if i is not None)
+  def get_arg_at_dim(i, dim, arg):
+    if dim is None:
+      # Broadcast unmapped argument
+      return arg
+    return lax.index_in_dim(arg, i, axis=dim, keepdims=False)
+  outs = []
+  for i in range(axis_size):
```
Comment (mattjj) [inline jax/_src/debugging.py:66]: We could consider generating a rolled loop, i.e. a lax.map. That might avoid inducing large compile times. Can do that in a follow-up though.

File: jax/_src/debugging.py [hunk 2]
```diff
@@ -48,11 +53,21 @@ def debug_callback_abstract_eval(*flat_avals, callback: Callable[..., Any],
...
-    effect: DebugEffect, in_tree: tree_util.PyTreeDef):
-  del flat_args, callback, effect, in_tree
-  # TODO(sharadmv): implement batching rule
-  raise NotImplementedError('Batching not supported for `debug_callback`.')
+def debug_callback_batching_rule(args, dims, **params):
+  """Unrolls the debug callback across the mapped axis."""
+  axis_size = next(x.shape[i] for x, i in zip(args, dims)
+                   if i is not None)
+  def get_arg_at_dim(i, dim, arg):
+    if dim is None:
```
Comment (mattjj) [inline jax/_src/debugging.py:61]:
```suggestion
    if dim is batching.not_mapped:
```

Doesn't matter much but someday we might want to replace the use of `None` with some other sentinel object.

File: jax/_src/debugging.py (hunk 1 above)
Comment (sharadmv) [reply jax/_src/debugging.py:66]: Added a TODO

**Review by mattjj — APPROVED** [review]
Beautiful!

File: jax/_src/debugging.py [hunk 3]
```diff
@@ -48,11 +53,23 @@ def debug_callback_abstract_eval(*flat_avals, callback: Callable[..., Any],
   del flat_avals, callback, in_tree
   return [], {effect}
 
-def debug_callback_batching_rule(*flat_args, callback: Callable[..., Any],
-    effect: DebugEffect, in_tree: tree_util.PyTreeDef):
-  del flat_args, callback, effect, in_tree
-  # TODO(sharadmv): implement batching rule
-  raise NotImplementedError('Batching not supported for `debug_callback`.')
+def debug_callback_batching_rule(args, dims, **params):
```
Comment (mattjj) [inline jax/_src/debugging.py:56]: James has made the point that this drops info about the axis name. But with custom batching rules around the corner, I think that gives the user a good way to change the default behavior if they want.

**Pull request merged**